metaci/testresults/models.py

```diff
@@ -274,3 +255,300 @@ class TestResultAsset(models.Model):
+            last_already_created = cls.objects.order_by("week_start").last()
+            if last_already_created:
+                start = last_already_created.week_start
+                # Let's always do at least one week in case it was incomplete
+                # (the week may not have been finished when we ran the scripts)
+                start = start - timezone.timedelta(days=1)
+
+            else:  # Starting from scratch
+                first_buildflow = (
+                    BuildFlow.objects.filter(time_end__isnull=False)
+                    .order_by("time_end")
+                    .first()
+                )
+                if first_buildflow:
+                    start = first_buildflow.time_end.date()
+                else:
+                    return None, None  # nothing to do if buildflows table is empty!
+
+        if end_string:
+            end = datetime.datetime.strptime(end_string, DATE_FORMAT).date()
+            end = end.replace(tzinfo=gettz())
+        else:
+            end = (
+                BuildFlow.objects.filter(time_end__isnull=False)
+                .order_by("-time_end")
+                .first()
+                .time_end.date()
+            )
+        return start, end
+
+    @classmethod
+    def date_range(cls, start_date, end_date):
+        """Generate a list of weeks described by two YYYY-MM-DD strings"""
+        start, end = cls._parse_start_and_date_dates(start_date, end_date)
+
+        # Turns out there is nothing to summarize
+        if not start:
+            return []
+
+        # start weeks on Sunday
+        start = cls._get_sunday(start)
+
+        dates_generator = (
+            start + datetime.timedelta(days=x)
+            for x in range(0, (end - start).days + 1, 7)
+        )
+
+        return dates_generator
+
+    @classmethod
+    def _get_sunday(cls, day):
+        """Go backwards to find the previous Sunday"""
+        day_of_week = (day.weekday() + 1) % 7  # Sunday is 0, Monday is 1, etc.
+        to_subtract = datetime.timedelta(days=day_of_week)
+        sunday_of_week = day - to_subtract
+        return sunday_of_week
+
+    @classmethod
+    def summarize_week(cls, date):
+        """Build summarize rows for a week containing a date"""
+        assert date
+        if not hasattr(date, "weekday"):
+            date = datetime.datetime.strptime(date, "%Y-%m-%d")
+
+        week_start = cls._get_sunday(date)
+        week_start_with_timezone = timezone.template_localtime(week_start, use_tz=True)
+        week_end_with_timezone = timezone.template_localtime(
+            week_start + timezone.timedelta(days=7), use_tz=True
+        )
+
+        method_contexts = cls._get_queryset_for_dates(
+            week_start_with_timezone,
+            week_end_with_timezone,
+            week_start=Value(week_start, output_field=models.DateField()),
+        )
+
+        obsolete_objects = cls.objects.filter(week_start=week_start)
```

Comment:
comment

Reply:
Part of 534c3404ab448173f73b8876f3d361b2642767bc